### PR TITLE
Updated logic to Not allow future date as incorporation date

### PR DESCRIPTION
**Pull request by PrashantMangukiya**
@puneetlath PR is ready for review.

### Details
Updated logic to show error whenever future date entered as incorporation date. 
So now **it will not allow to proceed if  future date entered** as company incorporation date.  
Proposal:   https://github.com/Expensify/App/issues/5722#issuecomment-945410229
Confirmation:  https://github.com/Expensify/App/issues/5722#issuecomment-946213677

### Fixed Issues
$ https://github.com/Expensify/App/issues/5722

### Tests | QA Steps
1. Launch the app
2. Log in with expensifail account
3. Enable Staging server
4. Click on deeplink https://staging.new.expensify.com/bank-account/contract
5. Log in with Chase credentials
6. In Company Information Page put future date as Incorporation Date. 
       **For example 01.01.2022 or Any date greater than current date.**
7. Click Save & Continue.
It will show error message "Please enter a valid date" under incorporation date if future date entered. 
**So It will not allow to proceed if future date entered as incorporation date.**

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

#### Web
https://user-images.githubusercontent.com/7823358/139823646-e9e92e8a-6009-4a98-8a65-0bf0ccf7c82c.mp4

#### Mobile Web
https://user-images.githubusercontent.com/7823358/139823751-02071d36-5dc0-431d-8360-aa4deb39052e.mov

#### Desktop
https://user-images.githubusercontent.com/7823358/139823868-2a2a5480-a490-4d8d-ace2-35fbabdf5468.mp4

#### iOS
https://user-images.githubusercontent.com/7823358/139823908-7164b75f-2044-48a9-8cb6-95938a3b7457.mov

#### Android
https://user-images.githubusercontent.com/7823358/139824118-d0d54197-1484-40f3-8246-5bb25bb456ee.mp4

### CODE REVIEW

File: src/libs/ValidationUtils.js
Comment (puneetlath) [inline src/libs/ValidationUtils.js:72]: What's the reason for adding `startOf('day')`?

**Review by puneetlath — CHANGES_REQUESTED** [review]

File: src/pages/ReimbursementAccount/CompanyStep.js
Comment (PrashantMangukiya) [reply src/pages/ReimbursementAccount/CompanyStep.js:156]: Yes we can do that. So there will be multiple error associated with incorporation date. We needs to update code as shown under. 

Within **src/pages/ReimbursementAccount/CompanyStep.js**  add/update code as under:
```javascript
constructor(props) { 
  ..,
  this.errorTranslationKeys = { 
      ....,
      incorporationDateFuture: 'bankAccount.error.incorporationDateFuture',
  } 
  this.clearIncorporationDateErrorsAndSetValue = this.clearIncorporationDateErrorsAndSetValue.bind(this);
}

/**
 * Clear errors associated with incorporation date.
 *
 * @param {String} value
 */
clearIncorporationDateErrorsAndSetValue(value) {
    this.clearErrorAndSetValue('incorporationDate', value);
    this.clearErrorAndSetValue('incorporationDateFuture', value);
}

/**
 * @returns {Boolean}
 */
validate() {
    .... 

    if (!isValidDate(this.state.incorporationDate)) {
        errors.incorporationDate = true;
    }

    if (!isValidPastDate(this.state.incorporationDate)) {
        errors.incorporationDateFuture = true;
    }
    ....
}

<DatePicker
    label={this.props.translate('companyStep.incorporationDate')}
    //onChange={value => this.clearErrorAndSetValue('incorporationDate', value)}   // ** OLD
    onChange={this.clearIncorporationDateErrorsAndSetValue}   // ** Updated
    value={this.state.incorporationDate}
    placeholder={this.props.translate('companyStep.incorporationDatePlaceholder')}
    // errorText={this.getErrorText('incorporationDate')}    // *** OLD
    errorText={this.getErrorText('incorporationDate') || this.getErrorText('incorporationDateFuture')}  // ** Updated
    translateX={-14}
/>
```

We need to add translation within en.js as under
```javascript
bankAccount: { 
   error: {
      ...
      incorporationDateFuture: 'Incorporation date cannot be in the future',
      ...
   }
}
```

We need to add translation within es.js as under
```javascript
bankAccount: { 
   error: {
      ...
      incorporationDateFuture: 'La fecha de incorporación no puede ser futura',
      ...
   }
}
```

**I did quick implementation and recorded screen, it looks as under:**

https://user-images.githubusercontent.com/7823358/137957998-a0c60c34-ed45-4315-8a63-8da9525d79cc.mov

Please let me know if we are planning to incorporate this within this issue, I can submit new commit within this pr.

Comment (PrashantMangukiya) [reply src/pages/ReimbursementAccount/CompanyStep.js:156]: Should I proceed to submit new commit within this pull request? I can send commit within few minutes If we are planning to proceed with this.

Comment (puneetlath) [reply src/pages/ReimbursementAccount/CompanyStep.js]: yes please!

**Review by parasharrajat — CHANGES_REQUESTED** [review]

File: src/pages/ReimbursementAccount/CompanyStep.js
Comment (Luke9389) [reply src/pages/ReimbursementAccount/CompanyStep.js:276]: Have you tested this? I don't think `this.getErrorText('incorperationDate')` will ever return a falsey value (pretty sure it'll always return the error text string, even if that error isn't happening).

Comment (PrashantMangukiya) [reply src/pages/ReimbursementAccount/CompanyStep.js:276]: Yes I tested and works as expected.  

Basically below line will return error text related to _incorporationDate_ or _incorporationDateFuture_
`errorText={this.getErrorText('incorporationDate') || this.getErrorText('incorporationDateFuture')}` So we do not need true or false value for this.

So if `this.getErrorText('incorporationDate')` return error text then it will show that error, otherwise it will show error text received from `this.getErrorText('incorporationDateFuture')`. 

e.g. Whenever entered date is more than thousand year from current date, it shows error text from this.getErrorText('incorperationDate') e.g. **Please enter a valid date**  (this logic was already there). 

e.g. Another case is entered date is future date (but not more than thousand year from now) so in that case it will show error message from `this.getErrorText('incorporationDateFuture')` e.g. **Incorporation date cannot be in the future**

I also recorded screen for this implementation, here it is: https://github.com/Expensify/App/pull/5939#discussion_r732077137  it is working as expected.

Comment (Luke9389) [reply src/pages/ReimbursementAccount/CompanyStep.js:276]: Ah, OK. I just looked at getErrorText and saw that it returns an empty string when the error isn't occurring. I think the naming here was confusing for me (I thought it would go look up the localized error text for a given error name).

File: src/pages/ReimbursementAccount/CompanyStep.js
Comment (Luke9389) [reply src/pages/ReimbursementAccount/CompanyStep.js:149]: `clearError` will clear multiple errors at once if you pass them. I think a better way to write this would be 

```js
clearDateErrorsAndSetValue(value) {
    clearError(['incorporationDate', 'incorporationDateFuture']);
    this.setValue(value);
}
```

Comment (PrashantMangukiya) [reply src/pages/ReimbursementAccount/CompanyStep.js:149]: Ok got it. i.e rename `clearIncorporationDateErrorsAndSetValue()` -->  `clearDateErrorsAndSetValue()`
and function code should be as under. ( I did small syntax correction)
``` javascript
clearDateErrorsAndSetValue(value) {
    this.clearError(['incorporationDate', 'incorporationDateFuture']);
    this.setValue({incorporationDate: value});
}
```
I also tested locally with this correction. It is working as expected. I will push new commit in few hours.

**Comment by PrashantMangukiya** [comment]
@puneetlath  can you please review the pr again, I also updated screen record with latest implementation. It shows one file conflict, I think it can be resolved before merge. Please let me know if anything need from my side. Thanks.

**Comment by puneetlath** [comment]
@PrashantMangukiya I'll give it another review! Can you please go ahead and resolve the merge conflict?

**Review by puneetlath — DISMISSED** [review]

**Comment by PrashantMangukiya** [comment]
@puneetlath thanks, 

this pull request shows me this message **Only those with write access to this repository can merge pull requests.**   It looks like I have no write access to this repo. Can you please guide me if any way to resolve conflict from my side ? I will be happy to do it. 
Screenshot 2021-11-02 at 10 27 58 PM

**Comment by puneetlath** [comment]
What do you see when you click the "resolve conflicts" button?

**Comment by PrashantMangukiya** [comment]
@puneetlath thanks for helping. It allowed me to edit code and resolve when clicked Resolve Conflicts button. I just resolved it. Let us wait to finish all checks.

**Comment by puneetlath** [comment]
Happy to help!

File: src/pages/ReimbursementAccount/CompanyStep.js
Comment (Luke9389) [reply src/pages/ReimbursementAccount/CompanyStep.js]: Ah, that was my bad, I was looking at the wrong method when I thought `clearError` could take multiple error keys.
Could you change it to this then? I think it's more readable and oblivious that the errors are different. 

```js
clearDateErrorsAndSetValue(value) {
        this.clearError('incorporationDate'); 
        this.clearError('incorporationDateFuture');
        this.setValue({incorporationDate: value});
}
```

Also would you change the comment on this method to: 
`Clear both errors associated with incorporation date, and set the new value.`

Comment (PrashantMangukiya) [reply src/pages/ReimbursementAccount/CompanyStep.js]: Yes this looks better. Updated it as per your request also pushed new commit.

**Comment by PrashantMangukiya** [comment]
@Luke9389  Updated `clearDateErrorsAndSetValue()` code as per your request, also pushed new commit. You can review it. Thanks.

**Review by puneetlath — DISMISSED** [review]

**Review by Luke9389 — DISMISSED** [review]
Thanks for the changes 👍
I ask every contributor this one last question after reviews are over. Have you tested this on all platforms after the most recent change? This helps us reduce regressions and often ends up in the contributor being paid faster. 

Also, we have a [HOLD] right now due to a company retreat, so we need to wait on merging this.

**Comment by PrashantMangukiya** [comment]
Yes, I did testing locally before every commit I pushed to pr, so you can freely merge it once hold released.

**Review by Luke9389 — APPROVED** [review]

**Review by puneetlath — APPROVED** [review]

**Comment by mallenexpensify** [comment]
Removed Company Offsite HOLD, feel free to merge @puneetlath

**Pull request merged**
